### PR TITLE
feat(client): add `refetchWhenOffline` option

### LINKED ISSUE #4935: Allow pausing the session polling when the browser is offline
### Description 📓

When the session polling is enabled (`refetchInterval` option), the request to `/session` is made regardless of the browser networks status. If the browser is offline (i.e. computer puts itself to sleep), the request fails, and the session is invalidated (`null`). This behavior can cause UI issues and make other requests fail when the application stores access tokens in the session object.

### Example (next-auth + SWR):
1. The app is mounted, and SWR fetches data successfully.
2. The host device goes offline => `/session` request fails => `session` is invalidated (null).
3. The host device goes back online, triggering the window focus event.
4. SWR tries to fetch the data but sends unauthenticated requests because the session is null, resulting in an error.

### Suggested solution:
We could add a new flag similar to the SWR's `refreshWhenOffline`. When disabled, all requests would be paused while the browser is offline, so the session would not be invalidated. This flag wouldn't introduce any breaking change, but it'd make the behavior configurable by the user.

The issue is related to #2215, but the stale bot didn't reopen the issue when I commented on it. As an alternative to this suggestion, we could stop the session from being invalidated, but I agree with the point made in https://github.com/nextauthjs/next-auth/issues/2215#issuecomment-867190636.

### How to reproduce ☕️

1. Run the [example app](https://codesandbox.io/s/next-auth-example-1kktb) with at least one configured provider (Auth0, for example). 
2. Ensure the session polling is enabled (`refetchInterval`).
3. Sign in with the previously configured auth provider.
4. Use the dev tools to set the throttling to "Offline".
5. Wait until the application tries to refetch the session.
6. See the session gets invalidated.

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

### PR DESCRIPTION
**Pull request by balazsorban44**
## ☕️ Reasoning

When the session polling is enabled (`refetchInterval` option), the request to `/api/auth/session` is made regardless of the browser network's status. If the browser is offline (i.e. computer puts itself to sleep), the request fails, and the session is invalidated (`null`). This behavior can cause UI issues and make other requests fail when the application stores access tokens in the session object.

This PR adds a new flag similar to the [SWR's `refreshWhenOffline`](https://swr.vercel.app/docs/options#options). When disabled, all requests will be paused while the browser is offline, so the session would not be invalidated. This flag does not introduce any breaking change, but it'd make the behavior configurable by the user.

In my opinion, it would make sense to flip this to be the default behavior in the next major version.

## 🧢 Checklist

- [x] Documentation
- [ ] Tests
- [x] Ready to be merged

## 🎫 Affected issues

Fixes #4935

## 📌 Resources

- [Contributing guidelines](./CONTRIBUTING.md)
- [Code of conduct](./CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

File: packages/next-auth/src/react/index.tsx [hunk 1]
```diff
@@ -404,13 +408,16 @@ export function SessionProvider(props: SessionProviderProps) {
     // Set up polling
     if (refetchInterval) {
       const refetchIntervalTimer = setInterval(() => {
-        if (__NEXTAUTH._session) {
+        // TODO: Flip this behavior in next major version
+        const shouldRefetch =
+          props.refetchWhenOffline !== false || navigator.onLine
```
Comment (ThangHuuVu) [inline packages/next-auth/src/react/index.tsx:413]: We could simplify this one a bit
```suggestion
          props.refetchWhenOffline || navigator.onLine
```

**Comment by viperfx** [comment]
Hi folks - is there any workaround for this or ETA on when this may go in? This is an issue we are hitting often as our endusers often are mobile and get their session invalidated when offline.

File: packages/next-auth/src/react/index.tsx [hunk 2]
```diff
@@ -404,13 +408,16 @@ export function SessionProvider(props: SessionProviderProps) {
     // Set up polling
     if (refetchInterval) {
       const refetchIntervalTimer = setInterval(() => {
-        if (__NEXTAUTH._session) {
+        // TODO: Flip this behavior in next major version
+        const shouldRefetch =
```
[the remark this answers is not in the document]
Comment (maieulChevalier) [reply packages/next-auth/src/react/index.tsx:412]: I don't know if your approach is better than balazsorban's, but I would rename your variables for better readability:

```javascript
function useNavigatorOnline() {
  const [isOnline, setIsOnline] = React.useState(typeof navigator !== 'undefined' ? navigator.onLine : false);

  React.useEffect(() => {
    window.addEventListener('online', setIsOnline(true));
    window.addEventListener('offline', setIsOnline(false);

    return () => {
      window.removeEventListener('online', setIsOnline(true));
      window.removeEventListener('offline', setIsOnline(false));
    };
  }, []);

  return isOnline;
}
```
Don't know if `useNavigatorOnline` can be renamed to `isOnline`? Think it is easier to read as well if this function's goal is only to check if it's online :p.

**Comment by maieulChevalier** [comment]
> Hi folks - is there any workaround for this or ETA on when this may go in? This is an issue we are hitting often as our endusers often are mobile and get their session invalidated when offline.

You can wrap your logic in an `if(navigator.onLine) { your logic }`. My app would redirect to the signin page when offline because it has a redirect based on user roles. Wrapping the redirect logic in this if statement solved it for me.

Here is an example: 

```jsx
export default function Layout({ children }) {
  const router = useRouter();
  const { data: session, status } = useSession();
  
  useEffect(() => {
    if (navigator.onLine) {
      if (status === "unauthenticated" || session === null) {
        router.push("/first-route");
      }
      if (status === "authenticated" && session?.user?.role === "role1") {
        router.push("/second-route");
      }
      if (status === "authenticated" && session?.user?.role === "role2") {
        router.push("/third-route");
      }
    }
        return;
  }, [status, session]);

  if (status === "loading") {
    return <Spinner />;
  }
  return (
    <>
      <main>{children}</main>
    </>
  );
}
```

File: packages/next-auth/src/react/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (ThangHuuVu) [reply packages/next-auth/src/react/index.tsx]: thanks for the suggestion @maieulChevalier , I applied

**Review by ThangHuuVu — APPROVED** [review]
Let's get this merged 🙌

**Comment by brendonco** [comment]
Which next-auth version to use with this changes?

**Comment by nik32** [comment]
<SessionProvider
  refetchInterval={60}
  refetchOnWindowFocus={typeof navigator !== "undefined" && navigator.onLine}
  refetchWhenOffline={false}
>
@ThangHuuVu @balazsorban44 hi guys! I tried this but it is not working. Session is still being refreshed even if user is offline

PS: the only thing which i did not configure was refetchInterval={60} as I don't want the network cost due to a call to session being made every 60secs (I have kept it as default i.e. 0)

**Comment by nik32** [comment]
@ThangHuuVu  @balazsorban44 Hi guys! This is not working. Kindly refer to my comment in https://github.com/nextauthjs/next-auth/issues/4935#issuecomment-1606108080

**Comment by fbenitezsura** [comment]
any solution ?

**Comment by AlexChika** [comment]
If you still face this issue, pls read my comments on the issue here  [issues/493](https://github.com/nextauthjs/next-auth/issues/4935#issuecomment-3243394652)

**Pull request merged**